### ISSUE TITLE
Config server requires restart for new remote repos

### ISSUE THREAD

**Issue by shakuzen** [spring-cloud/spring-cloud-config#265]
I am using Brixton.M2 and I noticed some behavior that was unexpected to me.

I started my config server, it was serving requests fine. Then I pushed a new branch to my git repo that the config server is using. I tried retrieving a plain text file (using the ResourceController's endpoint) with the new branch as the label. The config server cannot find the label and returns an error to that effect. I thought this was strange. I sent a POST request to the config server's `/refresh` endpoint thinking this might cause it to pull from the remote. However, it continues to not be able to find the new label until I restart the config server.

It would be nice if a full server restart were not required to serve newly pushed branches.

**Comment by dsyer** [comment]
Works for me. Can you provide more details (platform etc.)?

**Comment by shakuzen** [comment]
I was able to reproduce it on my Mac just now. It did not happen if you are pointing the config server at a local git repo (using `file:///`). Interestingly, after the first two requests failed with 'NoSuchLabelException: No such label: break', making a request for a branch it does know about (`master`) is successful, and then subsequent requests for the label `break` worked fine.

I made a basic Config Server from start.spring.io:
Spring Boot 1.3.2.RELEASE
Spring Cloud Brixton.BUILD-SNAPSHOT

`bootstrap.properties`
```properties
spring.cloud.config.server.git.uri=https://github.com/shakuzen/config-repo
spring.cloud.config.fail-fast=true
spring.cloud.config.server.bootstrap=true
encrypt.fail-on-error=false
```

**Steps**

* Point config server at remote git repo and start up config server
* Separately clone that repo
* Checkout a new branch
* Push that branch to the remote
* Hit a config server endpoint with that branch

**Expected:** non-error response
**Actual:** error response (`NoSuchLabelException: No such label: break`)

**Repo used:** https://github.com/shakuzen/config-repo

**Sample requests:**
```
$ curl -s 'http://localhost:8080/bar/default/break/static.conf' | jq .
{
  "timestamp": 1454507498091,
  "status": 500,
  "error": "Internal Server Error",
  "exception": "org.springframework.cloud.config.server.environment.NoSuchLabelException",
  "message": "No such label: break",
  "path": "/bar/default/break/static.conf"
}
```

```
$ curl -s 'http://localhost:8080/bar/default/break' | jq .
{
  "timestamp": 1454507510056,
  "status": 404,
  "error": "Not Found",
  "exception": "org.springframework.cloud.config.server.environment.NoSuchLabelException",
  "message": "No such label: break",
  "path": "/bar/default/break"
}
```

**Stacktrace:**
This is the stacktrace for the first request above. The second did not output a stacktrace because separate to this issue, `NoSuchLabelException` ends up as a 500 error from the ResourceController, but a 404 from the EnvironmentController.

```
org.springframework.cloud.config.server.environment.NoSuchLabelException: No such label: break
	at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.refresh(JGitEnvironmentRepository.java:155) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.getLocations(JGitEnvironmentRepository.java:121) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.getLocations(MultipleJGitEnvironmentRepository.java:111) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.environment.AbstractScmEnvironmentRepository.findOne(AbstractScmEnvironmentRepository.java:40) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.findOne(MultipleJGitEnvironmentRepository.java:140) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.environment.EnvironmentEncryptorEnvironmentRepository.findOne(EnvironmentEncryptorEnvironmentRepository.java:53) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
	at org.springframework.cloud.config.server.resource.ResourceController.resolve(ResourceController.java:74) ~[spring-cloud-config-server-1.1.0.BUILD-20160202.031501-91.jar:1.1.0.BUILD-SNAPSHOT]
```

**Comment by wbauern** [comment]
We are running into a similar type of problem.  We are using config server with an application placeholder like the following to use a different repository for each spring boot application:

```
spring:
  cloud:
    config:
      server:
        git:
          uri: ssh://git@git-server/{application}.git

```

If a git repository for the {application} is created on the git server before that spring boot app starts up and hits the config server through the endpoint to get its properties then everything works fine.

If a git repository for the {application} is not created first and the spring boot application starts up then it gets a "Cannot load configuration from ssh://git@git-server/fooapp.git". 

Now if we create the remote git repository for fooapp, add a fooapp.yml, commit it and push to the git-server we will continue to get the "Cannot load configuration from ssh://git@git-server/fooapp.git" messages and no properties from that file get loaded.  If the config server is restarted then everything works fine.

What we see happening behind the scenes is this:

When the config server endpoint is hit and a repository does exist for the {application} then the config server creates a local repository containing a .git/config with the following:

```
[core]
	symlinks = false
	repositoryformatversion = 0
	filemode = true
	logallrefupdates = true
	precomposeunicode = true
[remote "origin"]
	url = ssh://git@git-server/fooapp.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
	remote = origin
	merge = refs/heads/master
```

When the config server endpoint is hit and a remote repository does not exist for the {application} then the config server creates a local repository containing a .git/config with the following:

```
[core]
	symlinks = false
	repositoryformatversion = 0
	filemode = true
	logallrefupdates = true
	precomposeunicode = true
[remote "origin"]
	url = ssh://git@git-server/fooapp.git
	fetch = +refs/heads/*:refs/remotes/origin/*
```

It looks like it will never try to recreate the remote master branch after the first local repository is created unless the config server is restarted.   After tracing deeper into the code it looks like the "Cannot load configuration from ssh://git@git-server/fooapp.git" stems from a JGit exception thrown when it tries to get to the remote master branch.

An ugly work around we have discovered for this problem is to delete the local git repository that is missing the remote master branch then let it get recreated again the next time the config server's endpoint is hit.

Is this expected behavior?  If so, is there a more graceful workaround that you can think of?


Thanks,
Bill

**Comment by dsyer** [comment]
#439 is more specifically about new branches. Lets keep this one focused on new repos.

**Comment by haripriya-apptium** [comment]
I am unable to hit git repo afer configuring config server with application name and url with git repository .please help below is the strack trace of my issue.

Whitelabel Error Page

This application has no explicit mapping for /error, so you are seeing this as a fallback.

Sun Nov 13 10:18:06 EST 2016
There was an unexpected error (type=Internal Server Error, status=500).
Cannot clone or checkout repository

**Comment by ryanjbaxter** [comment]
@haripriya-apptium please dont randomly comment on issues.  Please open a new issue if you are having an issue, and provide as much information as possible so we can evaluate what might be wrong.

**Comment by dsyer** [comment]
This works for me now in 1.4.0 snapshots. I assume it was fixed as a side effect (maybe with #439).
